### PR TITLE
Eliminate allocation on a hot path with cyclic references

### LINKED ISSUE #47471: Find All References infinitely searches when Unity project Assembly-CSharp has a circular reference through assembly AOTGenerated
This issue tracks the bug report at https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1192910 (Microsoft-internal)

Apparently the compiler is going through a very slow path (which may include allocations) in `ReuseAssemblySymbols` when the project contains an indirect circular reference. 5GB allocations within code that doesn't accept a cancellation token (entry point is `CSharpCompilation.CommonAssembly`).

### PR DESCRIPTION
**Pull request by gafter**
Fixes #47471

Measurements indicate the principal problem in the reported scenario is an array allocation that
causes extreme garbage and results in GC thrashing.  That specific allocation is eliminated in this PR
by reusing a list as a buffer rather than allocating a new buffer for each call to `GetActualBoundReferencesUsedBy`.

Judging by the triply-nested loop in the caller (`ReuseAssemblySymbols`),
it is possible that some algorithmic work could further reduce any issues that remain.  However,
the customer's problem is already solved because they removed the circular reference in their
projects.  Thus this PR simply addresses the observed symptom.

There is no specific test for this as it would require a very large solution and measuring the
amount of memory allocation.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs [hunk 1]
```diff
@@ -124,24 +124,22 @@ protected override bool WeakIdentityPropertiesEquivalent(AssemblyIdentity identi
                 return AssemblyIdentityComparer.CultureComparer.Equals(identity1.CultureName, identity2.CultureName);
             }
 
-            protected override AssemblySymbol?[] GetActualBoundReferencesUsedBy(AssemblySymbol assemblySymbol)
+            protected override void GetActualBoundReferencesUsedBy(AssemblySymbol assemblySymbol, List<AssemblySymbol?> referencedAssemblySymbols)
             {
-                var refs = new List<AssemblySymbol?>();
+                referencedAssemblySymbols.Clear();
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs:129]: 💡 Consider making an empty list a precondition

**Review by sharwell — APPROVED** [review]

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs [hunk 2]
```diff
@@ -698,6 +698,9 @@ private void ReuseAssemblySymbols(BoundInputAssembly[] boundInputs, TAssemblySym
             Queue<AssemblyReferenceCandidate> candidatesToExamine = new Queue<AssemblyReferenceCandidate>();
             int totalAssemblies = assemblies.Length;
 
+            // A reusable buffer to contain the AssemblySymbols a candidate symbol refers to
+            List<TAssemblySymbol?> candidateReferencedSymbols = new List<TAssemblySymbol?>(1024);
```
Comment (sharwell) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs:702]:
```suggestion
            // A reusable buffer to contain the AssemblySymbols a candidate symbol refers to
            // ⚠ PERF: https://github.com/dotnet/roslyn/issues/47471
            List<TAssemblySymbol?> candidateReferencedSymbols = new List<TAssemblySymbol?>(1024);
```

**Comment by gafter** [comment]
May I please have a couple of reviews of this from the compiler team?

**Comment by jcouv** [comment]
@gafter Although I understand the PR doesn't include a test, I assume you were able to repro the issue locally. Was it just a matter of setting up mutually referring assemblies and using them in a compilation?

**Comment by sharwell** [comment]
@jcouv I walked through the proposed fix against trace and heap dump data gathered in the linked feedback ticket.

File: src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb [hunk 3]
```diff
@@ -52,21 +52,19 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 MyBase.New(simpleAssemblyName, identityComparer, observedMetadata)
             End Sub
 
-            Protected Overrides Function GetActualBoundReferencesUsedBy(assemblySymbol As AssemblySymbol) As AssemblySymbol()
-                Dim refs As New List(Of AssemblySymbol)
+            Protected Overrides Sub GetActualBoundReferencesUsedBy(assemblySymbol As AssemblySymbol, referencedAssemblySymbols As List(Of AssemblySymbol))
+                referencedAssemblySymbols.Clear()
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb:56]: On the VB side you clear the list, but on the C# side you assert that it's empty. Could we align the implementations?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4) with a question on VB vs. C# difference

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb (hunk 3 above)
Comment (jaredpar) [reply src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb:56]: Agree we should align this.

Comment (gafter) [reply src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb:56]: I missed this comment. Will follow up with a separate PR.

**Pull request merged**